Docs/source/usage/workflows/extend.rst

```diff
@@ -0,0 +1,229 @@
+
+   steps = 1000
+   for _ in range(steps):
+       sim.step(nsteps=1)
+
+       # do something custom with the sim object
+
+As a more flexible alternative, one can install `callback functions <https://en.wikipedia.org/wiki/Callback_(computer_programming)>`__, which will execute a given Python function at a
+specific location in the WarpX simulation loop.
+
+.. automodule:: pywarpx.callbacks
+   :members: installcallback, uninstallcallback, isinstalled
+
+
+pyAMReX
+-------
+
+Many of the following classes are provided through `pyAMReX <https://github.com/AMReX-Codes/pyamrex>`__.
+After the simulation is initialized, the pyAMReX module can be accessed via
+
+.. code-block:: python
+
+   from pywarpx import picmi, libwarpx
+
+   # ... simulation definition ...
+
+   # equivalent to
+   #   import amrex.space3d as amr
+   # for a 3D simulation
+   amr = libwarpx.amr  # picks the right 1d, 2d or 3d variant
+
+
+Full details for pyAMReX APIs are `documented here <https://pyamrex.readthedocs.io/en/latest/usage/api.html>`__.
+Important APIs include:
+
+* `amr.ParallelDescriptor <https://pyamrex.readthedocs.io/en/latest/usage/api.html#amrex.space3d.ParallelDescriptor.IOProcessor>`__: MPI-parallel rank information
+* `amr.MultiFab <https://pyamrex.readthedocs.io/en/latest/usage/api.html#amrex.space3d.MultiFab>`__: MPI-parallel field data
+* `amr.ParticleContainer_* <https://pyamrex.readthedocs.io/en/latest/usage/api.html#amrex.space3d.ParticleContainer_1_1_2_1_default>`: MPI-parallel particle data for a particle species
+
+
+Data Access
+-----------
+
+While the simulation is running, callbacks can read and write access the WarpX simulation data *in situ*.
+
+An important object in the ``pywarpx.picmi`` module for data access is ``Simulation.extension.warpx``, which is available only during the simulation run.
+This object is the Python equivalent to the C++ ``WarpX`` simulation class.
+
+.. py:class:: WarpX
+
+   .. py:method:: getistep(lev: int)
+
+      Get the current step on mesh-refinement level ``lev``.
+
+   .. py:method:: gett_new(lev: int)
+
+      Get the current physical time on mesh-refinement level ``lev``.
+
+   .. py:method:: getdt(lev: int)
+
+      Get the current physical time step size on mesh-refinement level ``lev``.
+
+   .. py:method:: multifab(multifab_name: str)
+
+      Return MultiFabs by name, e.g., ``"Efield_aux[x][l=0]"``, ``"Efield_cp[x][l=0]"``, ...
+
+      The physical fields in WarpX have the following naming:
+
+      - ``_fp`` are the "fine" patches, the regular resolution of a current mesh-refinement level
+      - ``_aux`` are temporary (auxiliar) patches at the same resolution as ``_fp``.
+        They usually include contributions from other levels and can be interpolated for gather routines of particles.
+      - ``_cp`` are "coarse" patches, at the same resolution (but not necessary values) as the ``_fp`` of ``level - 1``
+        (only for level 1 and higher).
+
+   .. py:method:: multi_particle_container
```

Comment:
```suggestion
   .. py:method:: multi_particle_container()
```
Add empty parenthesis to make it clear that this is a method.

Reply:
The empty parenthesis are removed for no-arg methods - this otherwise causes a warning in Spinx.

Reply:
Reference:
https://www.sphinx-doc.org/en/master/usage/domains/python.html#role-py-func

![image](https://github.com/ECP-WarpX/WarpX/assets/1353258/b45f444b-5b32-461c-88a1-dc97677c1002)

Reply:
Note that this will still *render* the `()` to the user in the manual.